vault/assumerolewithwebidentityprovider.go

```diff
@@ -0,0 +1,161 @@
+	}
+
+	req, resp := p.StsClient.AssumeRoleWithWebIdentityRequest(&sts.AssumeRoleWithWebIdentityInput{
+		RoleArn:          aws.String(p.RoleARN),
+		RoleSessionName:  aws.String(p.roleSessionName()),
+		DurationSeconds:  aws.Int64(int64(p.Duration.Seconds())),
+		WebIdentityToken: aws.String(webIdentityToken),
+	})
+	// Retry possibly temporary errors
+	req.RetryErrorCodes = append(req.RetryErrorCodes, sts.ErrCodeInvalidIdentityTokenException)
+
+	if err := req.Send(); err != nil {
+		return nil, err
+	}
+
+	log.Printf("Generated credentials %s using AssumeRoleWithWebIdentity, expires in %s", FormatKeyForDisplay(*resp.Credentials.AccessKeyId), time.Until(*resp.Credentials.Expiration).String())
+
+	return resp.Credentials, nil
+}
+
+func (p *AssumeRoleWithWebIdentityProvider) webIdentityToken() (string, error) {
+	const (
+		defaultMaxBufSize = 8192
+		defaultTimeout    = 2 * time.Minute
+	)
+
+	// Read OpenID Connect token from WebIdentityTokenFile
+	if p.WebIdentityTokenFile != "" {
+		b, err := ioutil.ReadFile(p.WebIdentityTokenFile)
+		if err != nil {
+			return "", fmt.Errorf("unable to read file at %s: %v", p.WebIdentityTokenFile, err)
+		}
+
+		return string(b), nil
+	}
+
+	// Exec WebIdentityTokenProcess to retrieve OpenID Connect token
+	var cmdArgs []string
+	if runtime.GOOS == "windows" {
+		cmdArgs = []string{"cmd.exe", "/C", p.WebIdentityTokenProcess}
+	} else {
+		cmdArgs = []string{"/bin/sh", "-c", p.WebIdentityTokenProcess}
+	}
+
+	r, w, err := os.Pipe()
+	if err != nil {
+		return "", err
+	}
+	defer r.Close()
+	defer w.Close()
+
+	cmd := exec.Command(cmdArgs[0], cmdArgs[1:]...)
+	cmd.Env = os.Environ()
+	cmd.Stdin = os.Stdin
+	cmd.Stdout = w
+	cmd.Stderr = os.Stderr
+
+	if err := cmd.Start(); err != nil {
+		return "", fmt.Errorf("failed to start command %q: %v", p.WebIdentityTokenProcess, err)
+	}
+	defer func() { _ = cmd.Process.Kill() }()
+
+	waitCh := make(chan error, 1)
+	go func() { waitCh <- cmd.Wait() }()
+
+	b := bytes.NewBuffer(make([]byte, 0, defaultMaxBufSize))
+	readCh := make(chan error, 1)
+	go func() {
+		w.Close() // close our write end of the pipe
+		defer r.Close()
+
+		_, err := io.CopyN(b, r, int64(defaultMaxBufSize))
+		readCh <- err
+	}()
+
+	timer := time.NewTimer(defaultTimeout)
+	defer timer.Stop()
+
+	// Wait for process to exit (or timeout)
+	select {
+	case err := <-waitCh: // process exited
+		if err != nil {
+			return "", fmt.Errorf("command %q failed: %v", p.WebIdentityTokenProcess, err)
+		}
+	case <-timer.C: // timeout
+		return "", fmt.Errorf("command %q timed out after %s", p.WebIdentityTokenProcess, defaultTimeout)
+	}
+
+	// Wait for read to finish (or timeout)
+	select {
+	case err := <-readCh: // process exited
+		if err != nil && err != io.EOF {
+			return "", fmt.Errorf("read output from %q failed: %v", p.WebIdentityTokenProcess, err)
+		}
+	case <-timer.C: // timeout
+		return "", fmt.Errorf("command %q timed out after %s", p.WebIdentityTokenProcess, defaultTimeout)
+	}
```

Comment:
There is a lot of code here with channels, defers, etc, and I'm wondering if this is more complex than it needs to be.

Would a `out, err := exec.Command(cmdArgs).Output()` suffice here?

Reply:
It would suffice for the simple case, yes. This bit of code is again inspired by [aws-sdk-go's implementation of processcreds](https://github.com/aws/aws-sdk-go/blob/0961961790e6dc3be58a7fa17bfc294143320536/aws/credentials/processcreds/provider.go#L332).

This code is more defensive than `exec.Command(cmdArgs).Output()` would be in some cases. Specifically:
* If the process writes more than `defaultMaxBufSize` bytes to stdout, we will stop reading and close our end of the pipe. The process will die with EPIPE/SIGPIPE. The amount of memory we use as a buffer is fixed. On the other hand, `Output` is implemented with a `bytes.Buffer` which will expand 'infinitely' (until it runs out of memory).
* It will timeout and kill the process after `defaultTimeout`, but this could be also achieved using `exec.CommandContext()`.

Since `aws-vault` is configured by the user for their own use on their own machine, we might not need to be as defensive to processes that run away and write lots to stdout. I think in the worse case, `aws-vault` would panic with `ErrTooLarge` and die? If you think that's the case, I'm happy to simplify this to use `exec.Command().Output` or `exec.ContextCommand().Output()`. Let me know :)

Reply:
Yes I don't think we need to be as defensive as the SDK, just go with the simple code I reckon

Reply:
Sounds good. I pushed that up as 78eb6646be3b96119337dce80955585c817c7e12.